<center>
<img src="../../img/ods_stickers.jpg">
## Открытый курс по машинному обучению. Сессия № 2
</center>
Автор материала: программист-исследователь Mail.ru Group, старший преподаватель Факультета Компьютерных Наук ВШЭ Юрий Кашницкий. Материал распространяется на условиях лицензии [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Можно использовать в любых целях (редактировать, поправлять и брать за основу), кроме коммерческих, но с обязательным упоминанием автора материала.

# <center> Домашнее задание № 1
## <center> Анализ данных по сердечно-сосудистым заболеваниям

**В задании предлагается с помощью Pandas ответить на несколько вопросов по данным о сердечно-сосудистых заболеваниях (качать данные не надо, они уже есть в репозитории). Данные использовались в соревновании [Ml Boot Camp 5](http://mlbootcamp.ru/round/12/sandbox/).**

**Заполните код в клетках (где написано "Ваш код здесь") и ответьте на вопросы в [веб-форме](https://goo.gl/8D8uDf).**

В соревновании предлагалось определить наличие/отсутствие сердечно-сосудистых заболеваний (ССЗ) по результатам осмотра пациента.

**Описание данных.**

Объективные признаки:

 - Возраст (age)
 - Рост (height)
 - Вес (weight)
 - Пол (gender)
 

Результаты измерения:

 - Артериальное давление верхнее и нижнее (ap_hi, ap_lo)
 - Холестерин (cholesterol)
 - Глюкоза (gluc)
 

Субъективные признаки (со слов пациентов):

 - Курение (smoke)
 - Употребление алкоголя (alco)
 - Физическая активность (active)
 
Целевой признак (который интересно будет прогнозировать):
 - Наличие сердечно-сосудистых заболеваний по результатам классического врачебного осмотра (cardio)

Значения показателей холестерина и глюкозы представлены одним из трех классов: норма, выше нормы, значительно выше нормы. Значения субъективных признаков — бинарны.

Все показатели даны на момент осмотра.

Мы будем работать только с обучающей выборкой и с помощью `Pandas` проведем первичный анализ данных. 

Из библиотек нам понадобятся только `NumPy` и `Pandas`.

In [1]:
import numpy as np
import pandas as pd

Считываем данные из CSV-файла в объект pandas DataFrame.  

In [2]:
df = pd.read_csv('../../data/mlbootcamp5_train.csv', sep=';', index_col='id')

Посмотрим не первые 5 записей.

In [3]:
df.head(7)


age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                          
0   18393       2     168    62.0    110     80            1     1      0   
1   20228       1     156    85.0    140     90            3     1      0   
2   18857       1     165    64.0    130     70            3     1      0   
3   17623       2     169    82.0    150    100            1     1      0   
4   17474       1     156    56.0    100     60            1     1      0   
8   21914       1     151    67.0    120     80            2     2      0   
9   22113       1     157    93.0    130     80            3     1      0   

    alco  active  cardio  
id                        
0      0       1       0  
1      0       1       1  
2      0       0       1  
3      0       1       1  
4      0       0       0  
8      0       0       0  
9      0       1       0

**<font color='red'>Вопрос 1 (1 балл).</font> Сколько мужчин и женщин представлено в этом наборе данных? Не было дано расшифровки признака "пол" (какому полу соответствует 1, а какому – 2 в признаке ``gender``) – это определите, посмотрев также на рост при разумном предположении, что в среднем мужчины выше.**

**<font color='red'>Варианты:</font>**
- **45530 женщин и 24470 мужчин**
- 45530 мужчин и 24470 женщин
- 45470 женщин и 24530 мужчин
- 45470 мужчин и 24530 женщин

In [4]:
# Ваш код здесь
# h_mean = df['height'].mean()
# def sex(height):
#     if height <= h_mean:
#         return 1
#     else:
#         return 0
    
# df['sex'] = df['height'].apply(sex)


# df['sex'] = df['sex'].astype('bool')

# df[df['sex'] == 0].describe()

print(df[df['gender'] == 1].describe())

print(df[df['gender'] == 2].describe()) 


                age   gender        height        weight         ap_hi  \
count  45530.000000  45530.0  45530.000000  45530.000000  45530.000000   
mean   19510.124577      1.0    161.355612     72.565605    128.139249   
std     2432.024493      0.0      7.052726     14.238838    158.054522   
min    10859.000000      1.0     55.000000     21.000000   -140.000000   
25%    18040.000000      1.0    157.000000     63.000000    120.000000   
50%    19726.000000      1.0    161.000000     70.000000    120.000000   
75%    21330.000000      1.0    166.000000     80.000000    140.000000   
max    23701.000000      1.0    250.000000    200.000000  16020.000000   

              ap_lo   cholesterol          gluc         smoke          alco  \
count  45530.000000  45530.000000  45530.000000  45530.000000  45530.000000   
mean      94.522776      1.384735      1.235054      0.017856      0.025500   
std      174.776855      0.693417      0.582956      0.132431      0.157639   
min      -70.0000

**<font color='red'>Вопрос 2 (1 балл).</font> Кто в среднем чаще указывает, что употребляет алкоголь – мужчины или женщины?**

**<font color='red'>Варианты:</font>**
- женщины
- **мужчины**

In [16]:
# Ваш код здесь

def sex(gender):
    if gender == 1:
         return 'female'
    else:
         return 'male'
    
df['sex'] = df['gender'].apply(sex)


#df['sex'] = df['sex'].astype('bool')

print(df[df['sex'] == 'male'].describe())

print(df[df['sex'] == 'female'].describe())


                age   gender        height        weight         ap_hi  \
count  24470.000000  24470.0  24470.000000  24470.000000  24470.000000   
mean   19392.097875      2.0    169.947895     77.257307    130.078872   
std     2529.752770      0.0      7.228895     14.187636    146.186152   
min    10798.000000      2.0     65.000000     10.000000   -150.000000   
25%    17502.000000      2.0    165.000000     68.000000    120.000000   
50%    19655.500000      2.0    170.000000     75.000000    120.000000   
75%    21321.000000      2.0    175.000000     85.000000    140.000000   
max    23713.000000      2.0    207.000000    200.000000  14020.000000   

              ap_lo   cholesterol          gluc         smoke          alco  \
count  24470.000000  24470.000000  24470.000000  24470.000000  24470.000000   
mean     100.551982      1.333633      1.210462      0.218880      0.106375   
std      211.557292      0.653764      0.551493      0.413496      0.308323   
min        0.0000

**<font color='red'>Вопрос 3 (1 балл).</font> Во сколько раз (округленно) процент курящих среди мужчин больше, чем процент курящих среди женщин (по крайней мере, по этим анкетным данным)?**

**<font color='red'>Варианты:</font>**
- 4
- **8**
- 12
- 16

In [50]:
# Ваш код здесь

# (((len(df[(df['smoke'] == 1) & (df['sex'] == 'male')])) / (len(df['sex'] == 'male'))))  / \
# (((len(df[(df['smoke'] == 1) & (df['sex'] == 'female')])) / (len(df['sex'] == 'female'))))  


print((((len(df[(df['smoke'] == 1) & (df['sex'] == 'male')])) / (len(df['sex'] == 'male')))) * 100) 

print((((len(df[(df['smoke'] == 1) & (df['sex'] == 'female')])) / (len(df['sex'] == 'female')))) * 100)  



7.651428571428571
1.1614285714285715


**<font color='red'>Вопрос 4 (1 балл).</font> Вы наверняка заметили, что значения возраста какие-то странные. Догадайтесь, в чем здесь измеряется возраст, и ответьте, на сколько месяцев (примерно) отличаются медианные значения возраста курящих и некурящих.**

**<font color='red'>Варианты:</font>**
- 5
- 10
- 15
- **20**

In [68]:
# Ваш код здесь

df.describe()

print(df[(df['age']) & (df['smoke'] == 1)]['age'].median()) 

print(df[(df['age']) & (df['smoke']) == 0]['age'].median())

(df[(df['age']) & (df['smoke']) == 0]['age'].median()) - (df[(df['age']) & (df['smoke'] == 1)]['age'].median())



19071.0
19714.0


643.0

**<font color='red'>Вопрос 5 (2 балла).</font> В [статье](https://ru.wikipedia.org/wiki/Сердечно-сосудистый_риск) на Википедии про сердечно-сосудистый риск показана шкала SCORE для расчёта риска смерти от сердечно-сосудистого заболевания в ближайшие 10 лет. Вот она:**
<img src='../../img/SCORE2007.png' width=70%>

Давайте посмотрим на правый верхний прямоугольник, отображающий сегмент курящих мужчин в возрасте от 60 до 65 лет. (Неочевидно, но тут для возраста и давления цифры означают верхнюю границу).

Видим 9-ку в левом нижнем углу этого прямоугольника и 47 – в правом верхнем. То есть если при этом систолическое (т.е. верхнее) артериальное давление – меньше 120 мм рт.ст., то риск ССЗ оценивается примерно в 5 раз ниже, чем если бы давление лежало в интервале [160, 180). 

Давайте посчитаем аналогичное значение, но на наших данных. 

Уточнения:
- Посчитайте признак ``age_years`` – возраст в годах, округлив до целых. Для данного примера отберите людей от 60 до 64 лет включительно
- Категории уровня холестрина на рисунке и в наших данных отличаются. Отображение значений на картинке в значения признака ``cholesterol`` следующее: 4 ммоль/л $\rightarrow$ 1, 5-7 ммоль/л $\rightarrow$ 2, 8 ммоль/л $\rightarrow$ 3.

Во сколько раз (округленно) отличаются доли больных в двух сегментах, описанных выше? Посчитайте на наших данных.

**<font color='red'>Варианты:</font>**
- 2
- 3
- 4
- 5

In [69]:
# Ваш код здесь

df.head()

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                          
0   18393       2     168    62.0    110     80            1     1      0   
1   20228       1     156    85.0    140     90            3     1      0   
2   18857       1     165    64.0    130     70            3     1      0   
3   17623       2     169    82.0    150    100            1     1      0   
4   17474       1     156    56.0    100     60            1     1      0   

    alco  active  cardio     sex  
id                                
0      0       1       0    male  
1      0       1       1  female  
2      0       0       1  female  
3      0       1       1    male  
4      0       0       0  female

**<font color='red'>Вопрос 6 (2 балла).</font> Постройте новый признак – BMI ([Body Mass Index](https://en.wikipedia.org/wiki/Body_mass_index)). Для этого надо вес в килограммах поделить на квадрат роста в метрах. Нормальными считаются значения BMI от 18.5 до 25. Выберите верные утверждения.**

**<font color='red'>Утверждения:</font>**
- Медианный BMI по выборке лежит в пределах нормы
- У женщин в среднем BMI выше, чем у мужчин
- У здоровых в среднем BMI выше, чем у больных
- В сегменте здоровых и непьющих мужчин BMI ближе к норме, чем в сегменте здоровых и непьющих женщин

In [9]:
# Ваш код здесь

**<font color='red'>Вопрос 7 (2 балла).</font> ** Можно заметить, что данные не особо-то чистые, много в них всякой "грязи" и неточностей. Еще лучше мы это увидим, когда обсудим визуализацию данных.

Отфильтруйте следующие сегменты пациентов (считаем это ошибками в данных)
- указанное нижнее значение артериального давления выше верхнего. 
- рост строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили (используте `pd.Series.quantile`, если не знаете, что это такое – прочитайте)
- вес строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили

Этот вовсе не вся чистка данных, которую можно было проделать, но пока остановимся на этом.

Сколько процентов данных (округленно) мы выбросили?

**<font color='red'>Варианты:</font>**
- 8
- 9
- 10
- 11

In [10]:
# Ваш код здесь